internal/auth/auth.go

```diff
@@ -0,0 +1,241 @@
+package auth
+
+import (
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/fabric8-services/fabric8-jenkins-proxy/internal/util"
+	"github.com/matryer/resync"
+	log "github.com/sirupsen/logrus"
+)
+
+// Client used to access auth service
+type Client struct {
+	URL          string
+	log          *log.Entry
+	publicKeys   sync.Map
+	updateWait   time.Duration // How long to wait after keys are fetched once
+	singleUpdate resync.Once
+}
+
+// TokenJSON represents a JSON Web Token
+type TokenJSON struct {
+	AccessToken      string `json:"access_token"`
+	RefreshToken     string `json:"refresh_token"`
+	TokenType        string `json:"token_type"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
+	Errors           []util.ErrorInfo
+}
+
+var defaultClient *Client
+
+// SetDefaultClient set auth client that will be returned by auth.DefaultClient
+func SetDefaultClient(c *Client) {
+	defaultClient = c
+}
+
+// DefaultClient returns default auth client
+func DefaultClient() *Client {
+	return defaultClient
+}
+
+// NewClient create a new auth client
+func NewClient(authURL string) *Client {
+	c := &Client{
+		URL:        authURL,
+		log:        log.WithField("component", "auth"),
+		updateWait: 5 * time.Minute,
+	}
+	go c.updatePublicKeysOnce()
+	return c
+}
+
+// UIDFromToken returns user identity on giving raw JWT token string
+func (c *Client) UIDFromToken(token string) (sub string, err error) {
+	t, err := jwt.Parse(token, c.publicKeyForToken)
+	if err != nil {
+		return
+	}
+
+	// to dump the token; uncomment below
+	// c.log.Debug("==================================================")
+	// c.log.Debugf("token: %s", spew.Sdump(t))
+	// c.log.Debug("==================================================")
+
+	claims, ok := t.Claims.(jwt.MapClaims)
+	if ok && t.Valid {
+		if claims["sub"].(string) == "" {
+			err = fmt.Errorf("Could not find user id in token")
+			return
+		}
+		sub = claims["sub"].(string)
+	}
+	return
+}
+
+// OSOTokenForCluster returns Openshift online token given the clusterURL and raw JWT token
+func (c *Client) OSOTokenForCluster(clusterURL, token string) (osoToken string, err error) {
```

Comment:
Shouldn't `token` be `accessToken`? Can you confirm that we are using raw JWT token here?

Reply:
no, it is expecting the raw token .. see below 

```go
	if len(tj.AccessToken) > 0 {
		osoToken = tj.AccessToken
	} else {
		err = fmt.Errorf("OSO access token empty for %s", c.URL)
	}
```

Reply:
@kishansagathiya no, you are right, it must be called AccessToken
so, I will change it to `accessToken`